### PR TITLE
Convert StateComponent to support functional colors internally

### PR DESCRIPTION
**Pull request by srt32**
supports https://github.com/primer/view_components/issues/232

Instead of green, red, purple we now support open, closed, merged.

We still support the old colors until functional colors have fully
landed into Primer CSS.

### CODE REVIEW

File: stories/primer/state_component_stories.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ class Primer::StateComponentStories < ViewComponent::Storybook::Stories
   story(:state) do
     controls do
       title "this is the title"
-      select(:color, Primer::StateComponent::COLOR_OPTIONS, :default)
+      select(:color, Primer::StateComponent::DEPRECATED_COLOR_MAPPINGS.keys, :default)
```
Comment (manuelpuyol) [inline stories/primer/state_component_stories.rb:9]:
```suggestion
      select(:color, Primer::StateComponent::COLOR_OPTIONS, :default)
```

no need to change this

File: app/components/primer/state_component.rb [hunk 2]
```diff
@@ -36,7 +43,7 @@ class StateComponent < Primer::Component
     #   <%= render(Primer::StateComponent.new(title: "title", size: :small)) { "Small" } %>
     #
     # @param title [String] `title` HTML attribute.
-    # @param color [Symbol] Background color. <%= one_of(Primer::StateComponent::COLOR_OPTIONS) %>
+    # @param color [Symbol] Background color. <%= one_of(Primer::StateComponent::DEPRECATED_COLOR_MAPPINGS.keys) %>
```
Comment (manuelpuyol) [inline app/components/primer/state_component.rb:46]:
```suggestion
    # @param color [Symbol] Background color. <%= one_of(Primer::StateComponent:: COLOR_OPTIONS) %>
```

File: app/components/primer/state_component.rb [hunk 3]
```diff
@@ -4,12 +4,19 @@ module Primer
   # Component for rendering the status of an item.
   class StateComponent < Primer::Component
     COLOR_DEFAULT = :default
-    COLOR_MAPPINGS = {
+    NEW_COLOR_MAPPINGS = {
+      open: "State--open",
+      closed: "State--closed",
+      merged: "State--merged"
```
Comment (manuelpuyol) [inline app/components/primer/state_component.rb:10]: I think the idea was to keep something like

```suggestion
      open: "State--green",
      closed: "State--red",
      merged: "State--purple"
```

and later on we can change `--green` to `--open`

Comment (srt32) [reply app/components/primer/state_component.rb:10]: this makes more sense.

File: test/components/state_component_test.rb [hunk 4]
```diff
@@ -62,4 +62,10 @@ def test_applies_additional_classes_to_the_underlying_element_when_given_custom_
 
     assert_selector(".State.State--red.State--small.additional.class-names.here")
   end
+
+  def test_supports_functional_colors
+    render_inline(Primer::StateComponent.new(title: "foo", color: :merged)) { "Merged" }
+
+    assert_selector(".State--merged")
```
Comment (srt32) [inline test/components/state_component_test.rb:69]:
```suggestion
    assert_selector(".State--purple")
```

File: app/components/primer/state_component.rb [hunk 5]
```diff
@@ -36,7 +43,7 @@ class StateComponent < Primer::Component
     #   <%= render(Primer::StateComponent.new(title: "title", size: :small)) { "Small" } %>
     #
     # @param title [String] `title` HTML attribute.
-    # @param color [Symbol] Background color. <%= one_of(Primer::StateComponent::COLOR_OPTIONS) %>
+    # @param color [Symbol] Background color. <%= one_of(Primer::StateComponent:: COLOR_OPTIONS) %>
```
Comment (manuelpuyol) [inline app/components/primer/state_component.rb:46]:
```suggestion
    # @param color [Symbol] Background color. <%= one_of(Primer::StateComponent::COLOR_OPTIONS) %>
```

oops, my bad

**Review by manuelpuyol — APPROVED** [review]

File: app/components/primer/state_component.rb [hunk 6]
```diff
@@ -4,12 +4,19 @@ module Primer
...
+      open: "State--green",
+      closed: "State--red",
+      merged: "State--purple"
+    }.freeze
+
+    DEPRECATED_COLOR_MAPPINGS = {
       COLOR_DEFAULT => "",
       :green => "State--green",
       :red => "State--red",
       :purple => "State--purple"
```
Comment (joelhawksley) [inline app/components/primer/state_component.rb:17]: I _believe_ this should work:

```suggestion
      open: "State--open",
      closed: "State--closed",
      merged: "State--merged"
    }.freeze

    DEPRECATED_COLOR_MAPPINGS = {
      COLOR_DEFAULT => "",
      green: "State--open",
      red: "State--closed",
      purple: "State--merged"
```

Doing so enables us to fully switch to the functional class names while migrating the consumers of this component to use the functional color names.

File: app/components/primer/state_component.rb [hunk 7]
```diff
@@ -4,12 +4,19 @@ module Primer
...
     }.freeze
+    COLOR_MAPPINGS = NEW_COLOR_MAPPINGS.merge(DEPRECATED_COLOR_MAPPINGS)
     COLOR_OPTIONS = COLOR_MAPPINGS.keys
 
     SIZE_DEFAULT = :default
```
Comment (joelhawksley) [inline app/components/primer/state_component.rb:22]: Can we also change the examples to use functional colors?

File: docs/content/components/state.md [hunk 8]
```diff
@@ -43,7 +43,7 @@ Component for rendering the status of an item.
 | Name | Type | Default | Description |
 | :- | :- | :- | :- |
 | `title` | `String` | N/A | `title` HTML attribute. |
-| `color` | `Symbol` | `:default` | Background color. One of `:default`, `:green`, `:red`, or `:purple`. |
+| `color` | `Symbol` | `:default` | Background color. One of `:open`, `:closed`, `:merged`, `:default`, `:green`, `:red`, or `:purple`. |
```
Comment (joelhawksley) [inline docs/content/components/state.md:46]: It's a little odd to have this argument named `color`, as that's also a system argument. But perhaps that's a concern for another time 😄

File: app/components/primer/state_component.rb (hunk 6 above)
Comment (srt32) [reply app/components/primer/state_component.rb:17]: When I did this locally, the doc examples rendered without color, which told me this was broken. Am I misunderstanding?

File: docs/content/components/state.md (hunk 8 above)
Comment (manuelpuyol) [reply docs/content/components/state.md:46]: I think this is part of the conversation about splitting `system_arguments` into different things

File: app/components/primer/state_component.rb (hunk 6 above)
Comment (manuelpuyol) [reply app/components/primer/state_component.rb:17]: The problem is that those classes aren't officially released by PrimerCSS and will only work in dotcom.
I think mapping the functional keys to color classes (like `open: "State--green"`) is fine by now. Once Primer releases the new classes we can migrate the classes (without impacting users!)

File: app/components/primer/state_component.rb (hunk 7 above)
Comment (srt32) [reply app/components/primer/state_component.rb:22]: Done in ce30e41e45768ee96550ddf8766290cbcc5ad2ff!

File: app/components/primer/state_component.rb (hunk 6 above)
Comment (srt32) [reply app/components/primer/state_component.rb:17]: I think we're saying we're go leave this change as is. @joelhawksley is this making sense to you?

Comment (joelhawksley) [reply app/components/primer/state_component.rb:17]: Ah, I see what's going on here.

The docs examples load Primer from a public CDN, not using the version we serve in dotcom. 

We _really_ need to update the PVC docs site to use the version of Primer CSS we're using on dotcom. Perhaps we could prioritize that ahead of this PR?

Comment (srt32) [reply app/components/primer/state_component.rb:17]: cool, will follow up in https://github.com/primer/view_components/issues/276

Comment (manuelpuyol) [reply app/components/primer/state_component.rb:17]: @srt32 now I think this change is valid :)

File: test/components/state_component_test.rb [hunk 9]
```diff
@@ -62,4 +62,10 @@ def test_applies_additional_classes_to_the_underlying_element_when_given_custom_
 
     assert_selector(".State.State--red.State--small.additional.class-names.here")
   end
+
+  def test_supports_functional_colors
+    render_inline(Primer::StateComponent.new(title: "foo", color: :merged)) { "Merged" }
+
+    assert_selector(".State--purple")
```
Comment (srt32) [inline test/components/state_component_test.rb:69]:
```suggestion
    assert_selector(".State--merged")
```

File: app/components/primer/state_component.rb [hunk 10]
```diff
@@ -4,12 +4,19 @@ module Primer
...
+    }.freeze
+
+    DEPRECATED_COLOR_MAPPINGS = {
       COLOR_DEFAULT => "",
-      :green => "State--green",
-      :red => "State--red",
-      :purple => "State--purple"
+      green: "State--open",
+      red: "State--closed",
+      purple: "State--merged"
```
Comment (srt32) [inline app/components/primer/state_component.rb:17]:
```suggestion
      :green => "State--open",
      :red => "State--closed",
      :purple => "State--merged"
```

linting strikes again

File: test/components/state_component_test.rb [hunk 11]
```diff
@@ -62,4 +62,10 @@ def test_applies_additional_classes_to_the_underlying_element_when_given_custom_
 
     assert_selector(".State.State--red.State--small.additional.class-names.here")
```
Comment (srt32) [inline test/components/state_component_test.rb:63]:
```suggestion
    assert_selector(".State.State--closed.State--small.additional.class-names.here")
```

**Comment by srt32** [comment]
This PR has me thinking..... given we don't bundle primer/css with this library, if consumers of this gem are consuming the old version of primer css, then this change would be breaking since they don't have the new functional classes defined yet. How do we want to approach this?

**Comment by joelhawksley** [comment]
> This PR has me thinking..... given we don't bundle primer/css with this library, if consumers of this gem are consuming the old version of primer css, then this change would be breaking since they don't have the new functional classes defined yet. How do we want to approach this?

@srt32 good point. There are only 4 other public users of the library: https://github.com/primer/view_components/network/dependents?package_id=UGFja2FnZS0xMzY1MzYwNjkz, so I'm tempted not to worry about it too much. I do think we should point this out in the changelog for the release, as the first entry.

It might also be worth adding a note about the implicit dependency to the `primer.style/view-components` landing page 🤔

**Comment by srt32** [comment]
Is css-next as a private repo a typical workflow we do? Ideally we could point at a version of primer/css that anyone could consume.

**Comment by srt32** [comment]
We also do call out the dependency already ;)

https://primer.style/view-components/#dependencies

File: CHANGELOG.md [hunk 12]
```diff
@@ -2,6 +2,8 @@
 
 ## main
 
+* **Breaking change**: Upgrade StateComponent to support functional colors. This change requires using https://www.npmjs.com/package/@primer/css-next.
```
Comment (srt32) [inline CHANGELOG.md:5]: I'm not sure how best to communicate this. We need another sentence saying what's next.

File: CHANGELOG.md [hunk 13]
```diff
@@ -2,6 +2,8 @@
 
 ## main
 
+* **Breaking change**: Upgrade `StateComponent` to support functional colors. This change requires using https://www.npmjs.com/package/@primer/css-next. The required changes will be upstreamed to @primer/css at a later date.
```
Comment (srt32) [inline CHANGELOG.md:5]:
```suggestion
* **Breaking change**: Upgrade `StateComponent` to support functional colors. This change requires using [@primer/css-nex](https://www.npmjs.com/package/@primer/css-next). The required changes will be upstreamed to @primer/css at a later date. 
```

File: CHANGELOG.md [hunk 14]
```diff
@@ -2,6 +2,8 @@
 
 ## main
 
+* **Breaking change**: Upgrade `StateComponent` to support functional colors. This change requires using [@primer/css-nex](https://www.npmjs.com/package/@primer/css-next). The required changes will be upstreamed to @primer/css at a later date.
```
Comment (srt32) [inline CHANGELOG.md:5]:
```suggestion
* **Breaking change**: Upgrade `StateComponent` to support functional colors. This change requires using [@primer/css-next](https://www.npmjs.com/package/@primer/css-next). The required changes will be upstreamed to @primer/css at a later date. 
```

**Review by manuelpuyol — APPROVED** [review]

File: CHANGELOG.md [hunk 15]
```diff
@@ -2,6 +2,8 @@
 
 ## main
 
+* **Breaking change**: Upgrade `StateComponent` to support functional colors. This change requires using [@primer/css-next](https://www.npmjs.com/package/@primer/css-next). The required changes will be upstreamed to @primer/css at a later date.
```
Comment (srt32) [inline CHANGELOG.md:5]:
```suggestion
* **Breaking change**: Upgrade `StateComponent` to support functional colors. This change requires using [@primer/css-next](https://www.npmjs.com/package/@primer/css-next). The required changes will be upstreamed to @primer/css at a later date.
```

**Pull request merged**